alma/lib/Model/AddressData.php

```diff
@@ -0,0 +1,177 @@
+
+    /** @var string Country of shipping */
+    private $countryShipping;
+
+    /** @var string Country of billing */
+    private $countryBilling;
+
+    /** @var string|null State Province of shippîng */
+    private $shippingStateProvinceName;
+
+    /** @var string|null State Province of billing */
+    private $billingStateProvinceName;
+
+    /**
+     * Address Data construct
+     *
+     * @param Cart $cart
+     */
+    public function __construct(
+        Cart $cart
+    ) {
+        $this->cart = $cart;
+    }
+
+    /**
+     * Get shipping address of cart
+     *
+     * @return Address
+     */
+    public function getShippingAddress()
+    {
+        if (!$this->shippingAddress) {
+            $this->shippingAddress = new Address((int) $this->cart->id_address_delivery);
+        }
+
+        return $this->shippingAddress;
+    }
+
+    /**
+     * Get billing address of cart
+     *
+     * @return Address
+     */
+    public function getBillingAddress()
+    {
+        if (!$this->billingAddress) {
+            $this->billingAddress = new Address((int) $this->cart->id_address_invoice);
+        }
+
+        return $this->billingAddress;
+    }
+
+    /**
+     * Get country shipping address
+     *
+     * @return string
+     */
+    public function getShippingCountry()
+    {
+        if (!$this->countryShipping) {
+            $this->countryShipping = Country::getIsoById($this->getShippingAddress()->id_country);
+        }
+
+        return $this->countryShipping;
+    }
+
+    /**
+     * Get country billing address
+     *
+     * @return string
+     */
+    public function getBillingCountry()
+    {
+        if (!$this->countryBilling) {
+            $this->countryBilling = Country::getIsoById($this->getBillingAddress()->id_country);
+        }
+
+        return $this->countryBilling;
+    }
+
+    /**
+     * Get state of province shipping
+     *
+     * @return string|null
+     */
+    public function getShippingStateProvinceName()
+    {
+        if (is_null($this->shippingStateProvinceName) && $this->getShippingAddress()->id_state > 0) {
+            $this->shippingStateProvinceName = State::getNameById((int) $this->getShippingAddress()->id_state);
+        }
+
+        return $this->shippingStateProvinceName;
+    }
+
+    /**
+     * Get state of province billing
+     *
+     * @return string|null
+     */
+    public function getBillingStateProvinceName()
+    {
+        if (is_null($this->billingStateProvinceName) && $this->getBillingAddress()->id_state > 0) {
+            $this->billingStateProvinceName = State::getNameById((int) $this->getBillingAddress()->id_state);
+        }
+
+        return $this->billingStateProvinceName;
+    }
+
+    /**
+     * get phone addresses customer
+     *
+     * @return string
+     */
+    public function getShippingPhone()
+    {
+        $phone = null;
+        $shippingAddress = $this->getShippingAddress();
+
+        if ($shippingAddress->phone) {
```

Comment:
Stop the loop as soon as you find a phone instead of looping until the latest phone found

Reply:
There is no looping here....